README
ONTO-PERL suite
===============

ONTO-PERL a collection of PERL modules for dealing with any OBO-formatted
ontology (such as the Gene Ontology). This code distribution gathers 
object-oriented modules (for dealing with ontology elements such as 
Term, Relationship, Synonym and so forth), scripts (for typical tasks 
such as format conversions: obo2rdf, rdf2obo; ONTO-perl provides an 
extensive documentation as well as several examples that can be easily 
adapted for specific applications that involve ontology processing), and 
a set of test files to ensure the suite's implementation quality.

See also:

  1. perldoc ONTO-PERL.pod
  
  2. ONTO-PERL: An API supporting the development and analysis of bio-ontologies
     Antezana E, Egana M, De Baets B, Kuiper M, Mironov V.
     Bioinformatics (2008) 24 (6): 885-887; doi: 10.1093/bioinformatics/btn042

  3. ONTO-Toolkit: A set of Galaxy tools to deal with OBO-formatted ontologies,
     Antezana E, Venkatesan A, Mungall C, Mironov V, Kuiper M.
     BMC Bioinformatics 2010, 11(Suppl 12):S8 doi:10.1186/1471-2105-11-S12-S8

Visit the following URLs:

 - http://search.cpan.org/~easr/ONTO-PERL/

For installing this collection of modules:

	see the file INSTALL

or follow the standard CPAN procedure:

type the following mantra:

	perl -w Makefile.PL
	make
	make test
	make install

Other useful commands:

	make clean	
	make distcheck

DEPENDENCIES
============

ONTO-PERL requires the following modules:

1. Date::Manip (> v6.42):

	http://search.cpan.org/dist/Date-Manip/ (MANDATORY)

2. XML::Parser (> v2.34) is used in format conversions (obo2owl and owl2obo)

	http://search.cpan.org/~msergeant/XML-Parser/ (OPTIONAL)
	
2. XML::XPath (> v1.13) is used by the IntAct related modules:

   http://search.cpan.org/~msergeant/XML-XPath/ (OPTIONAL)

4. SWISS Knife (> v1.67) is used by the 'UniProtParser':

	ftp://ftp.ebi.ac.uk/pub/software/swissprot/Swissknife (OPTIONAL)	

5. Text::CSV (> v1.32) is used by 'bioportal_csv2obo.pl':

   http://search.cpan.org/~makamaka/Text-CSV/ (OPTIONAL)

COPYRIGHT AND LICENCE
=====================

Copyright (c) 2006-2014 by Erick Antezana. All rights reserved.
This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

Changes
Revision history for ONTO-PERL
==============================

0.01  Wed Jun  7 13:05:23 2006
	- original version; created by h2xs 1.23 with options
		-AXc -n CCO
		
0.xx ...

0.22 Improvements in exports

0.23 obo2owl methods were enhanced

0.24 Ontology.pm has many more functionalities

0.25 GOA parser, OBO parser enhanced

0.26 Enhanced: NCBIParser.pm, GoaParser.pm and Ontology.pm

0.27 obo2owl transformation updated, new modules for dealing with IDs

0.28 Performance improvement (code tuning)

0.29 More adjustments to follow OBO 1.2 spec, Set data structures improved

0.30 In sync with oboInOwl, IntAct modules

0.31 oboInOwl as well as ontology exports to OBO and OWL were improved

0.32 obo2owl and owl2obo scripts updated and improved, OBO and OWL Parsers enhanced

1.0  Ontology union/intersection methods are now available. Many more tests were added and examples of usage

1.01 CPAN test report fixed (SKIP tests when the required modules are missing)

1.02 'subsets' for relationship types are now supported, test improvements

1.03 export order improved (alphabetical way)

1.04 more 'order' improvements in test files

1.05 dates treatment is now WindowsTM compatible

1.06 documentation has been enhanced: many more examples, PODs, etc.

1.07 dbxref's parsing has been relaxed and improved. 'onto-perl' became 'ONTO-PERL' and some files were moved to more meaningful directories

1.08 RDF export is now available (obo2rdf.pl). Improved documentation.

1.09 Minor documentation improvements. Performance improvement for CCO-related modules. 

1.10 Added missing file (obo2rdf.pl)

1.11 Documentation and cosmetic improvements. Minor bug fixes.

1.12 CCO related modules were improved.

1.13 IntAct parser tests fixed. New RDF related modules.

1.14 Fixed issue related to the missing 'lib' directory.

1.15 Fixed issue with URL's that were hard-coded.

1.16 obo2owl ready for Galaxy.

1.17 Added some more scripts.

1.18 Added missing scripts.

1.19 Fixed old bug (multiple relationships between 2 terms). Added 3 more scripts (trimming, get sub-ontology, and get terms by name).

1.20 Fixed DbXref bug (http-like URLs).

1.21 New scripts (get_obsolete_terms.pl and get_root_terms.pl). POD improvements.

1.22 Added get_term_synonyms.pl; Added $VERSION to Ontology.pm; RegExp improvements.

1.23 Added support for 'created_by' and 'creation_date' tags; fixed bug in synonyms parsing.

1.24 Added new classes to deal with OBO ID in a more generic way; added many more tests; fixed bug when a term def was empty.

1.25 IDspaces are now better supported; more tests added; removed unnecessary code.

1.26 IDspaces bug fix; new set: IDspaceSet; code cleaning. Improved exports.

1.27 Closed some TODO's. Some refactoring to comply with the OBO guide. New modules Subset.pm and Map.pm

1.28 Added missing file (SubsetDefSet.pm), which broke many tests. Fixed minor bugs.

1.29 Added missing script (get_terms_and_synonyms.pl), minor bug fixes.

1.30 Code cleaning. Code optimisation, minor bug fixes.

1.31 Croak fixes. CCO modules clean-up.

1.32 Code clean up. Better regexp's. A few new sub's and tests. A few bug fixes. OBO::Core::Ontology->subset_def_set() became OBO::Core::Ontology->subset_def_map()

1.33 CCO modules updated. Minor improvements (e.g. get_relationships_by_(source|target)_term). New modules for instances. Relationship types are added while adding relationship.

1.34 Transitive reduction of ontologies implemented (OBO::Util::Ontolome). Minor performance enhancements. 

1.35 Term and Relationship type synonyms can be updated (based on the same synonym type name if defined). Fixed bug in Synonym.pm. Improved PODs. Dates in 'creation_date' entries are checked. Stricter module versions.

1.36 Term Local Neighbourhood retrieval implemented. Now using croak's instead of die's. Renaming: CCO --> APO. Checks in OBO::Core::Ontology::export() improved. New script for LCA. Bug fix in OBO::Util::SynonymSet::equals(). Some parsers were moved and cleaned.

1.37 Added missing data test file (up.dat). OBO::Core::Ontology::(delete_term()|set_term_id()) improved. Reflexive is_a relationships are now warned and skipped while parsing OBO files. SwissProtToRDF.pm added.

1.38 OBO::Core::Ontology::export is been "cleaned" and modularised (export2obo|rdf|owl|dot|gml|xml). File handles are better in test files. Parsers were improved. 'vis' exports are no longer supported. SwissProtToRDF.t added.

1.39 export2rdf bug fixed; SwissProtToRDF.t now checks if SWISS/Entry.pm is available. OBO::Core::(Instance|Relationship) were enhanced. Initial code to handle 'property_value' tags with instances. Fixed bug in Ontolome::transitive_reduction(). Experimental relationships composition implemented (for is_a and part_of).

1.40 Bug fixed in the parsing of the synonym scope. 'property_value' for OBO::Core::Term.

1.41 property_value code improved. dbxref bug fixed (URLs). UTF-8 support. onto-perl now understands 'default_relationship_id_prefix' and 'treat-xrefs-as-is_a'. All 'scripts (*.pl)' were cleaned. Relationship ID bug fixed. Reporting line counting bug fixed.

1.42 Makefile.PL improved. Subsets for Terms and Instances are now sorted in exports. Unused subsets and synonym type definitions: checked and removed while parsing. Refactoring: is_functional, is_inverse_functional. Check of recommended XML-parser primitives. Bug fix in Ontolome::transitive_closure().

1.43 Makefile.PL adapted: required modules, versions. Improvements in Ontolome::transitive_closure() and Ontology::export2rdf(). New script: bioportal_csv2obo.pl

1.43 Added missing file (bioportal_csv2obo.pl). Update of README and INSTALL.

Makefile.PL
# $Id: Makefile.PL 2014-01-17 09:22:13Z erick.antezana $
use 5.006;
use ExtUtils::MakeMaker;
use strict;

my $p = 'Date::Manip';
if (!eval "require $p") {
    print <<MSG;

   Date::Manip is missing. This module is required in order to perform checks 
   of dates while parsing OBO-formatted files (e.g. 'creation_date' tag).

MSG
      
}

$p = 'SWISS::Entry';
if (!eval "require $p") {
    print <<MSG;

   SWISS::Entry is missing. This module is only required in order to
   perform parsing operations with UniProt files. However, this 
   module is not required for using the rest of ONTO-PERL. If you do 
   not intend to use UniProt files, then you may continue, and install 
   SWISS::Entry in the future.

MSG
      
}

$p = 'XML::Parser';
if (!eval "require $p") {
    print <<MSG;

   XML::Parser is missing. This module is only required in order to
   perform parsing operations with XML files (obo2owl). However, this 
   module is not fully required for using the rest of ONTO-PERL. If you 
   do not intend to convert your files to OWL, then you may continue, 
   and install XML::Parser in the future.

MSG
      
}

$p = 'XML::XPath';
if (!eval "require $p") {
    print <<MSG;

   XML::XPath is missing. This module is only required in order to
   perform parsing operations with IntAct-XML files. However, this 
   module is not required for using the rest of ONTO-PERL. If you do 
   not intend to use IntAct XML files, then you may continue, and 
   install XML::XPath in the future.

MSG
      
}

$p = 'Text::CSV';
if (!eval "require $p") {
    print <<MSG;

   Text::CSV is missing. This module is only required in order to
   generate an OBO-formatted file from any CSV file found at the 
   BioPortal (http://bioportal.bioontology.org/). However, this 
   module is not required for using the rest of ONTO-PERL. If you do 
   not intend to use those CSV files, then you may continue, and 
   install Text::CSV in the future.

MSG
      
}

my @license = $ExtUtils::MakeMaker::VERSION > 6.84
? qw(LICENSE perl)
: ();

my @authors = ('Erick Antezana <erick.antezana -@- gmail.com>, Vladimir Mironov <vladimir.n.mironov -@- gmail.com>');

WriteMakefile(
	NAME               => 'Ontology',
	DISTNAME           => 'ONTO-PERL',
	ABSTRACT           => 'PERL modules for manipulating OBO-formatted ontologies, such as the Gene Ontology (GO)',
	VERSION_FROM       => 'lib/OBO/Core/Ontology.pm',
	MIN_PERL_VERSION   => '5.005',
	@license,
	PREREQ_PM          => {
						'Date::Manip'         => 6.42,
						#'SWISS::Entry'       => 1.67, # optional
						#'XML::Parser'        => 2.41, # optional
						#'XML::XPath'         => 1.13, # optional
						#'Text::CSV'          => 1.32  # optional
						},
	PREREQ_PRINT       => 'TRUE',
	BUILD_REQUIRES     => { 'Date::Manip' => 6.42, 'ExtUtils::MakeMaker' => 6.84 },
	CONFIGURE_REQUIRES => { 'Date::Manip' => 6.42, 'ExtUtils::MakeMaker' => 6.84 },
	EXE_FILES          => [
							qw(
								scripts/bioportal_csv2obo.pl
								scripts/get_ancestor_terms.pl
								scripts/get_child_terms.pl
								scripts/get_descendent_terms.pl
								scripts/get_lowest_common_ancestor.pl
								scripts/get_obsolete_term_id_vs_def_in_go.pl
								scripts/get_obsolete_term_id_vs_name_in_go.pl
								scripts/get_obsolete_terms.pl
								scripts/get_parent_terms.pl
								scripts/get_relationship_id_vs_relationship_def.pl
								scripts/get_relationship_id_vs_relationship_name.pl
								scripts/get_relationship_id_vs_relationship_namespace.pl
								scripts/get_relationship_types.pl
								scripts/get_root_terms.pl
								scripts/get_subontology_from.pl
								scripts/get_term_id_vs_term_def.pl
								scripts/get_term_id_vs_term_name.pl
								scripts/get_term_id_vs_term_namespace.pl
								scripts/get_term_local_neighbourhood.pl
								scripts/get_term_synonyms.pl
								scripts/get_terms.pl
								scripts/get_terms_and_synonyms.pl
								scripts/get_terms_by_name.pl
								scripts/go2owl.pl
								scripts/goa2rdf.pl
								scripts/obo2apo.pl
								scripts/obo2owl.pl
								scripts/obo2rdf.pl
								scripts/obo2tran.pl	
								scripts/obo2xml.pl
								scripts/obo_intersection.pl
								scripts/obo_transitive_reduction.pl
								scripts/obo_trimming.pl
								scripts/obo_union.pl
								scripts/owl2obo.pl
							)
						],
			
	($] >= 5.005 ?     ## Add these new keywords supported since 5.005
	(AUTHOR   => @authors) : ()),
	dist      => { 
					'SUFFIX'       => '.gz',
					'DIST_DEFAULT' => 'all tardist',
					'COMPRESS'     => 'gzip -9f' 
				},
	clean      => {FILES =>"t/data/test0.obo
							t/data/test1.obo
							t/data/test1.rdf
							t/data/test1.xml
							t/data/test2.dot
							t/data/test2.obo
							t/data/test2.owl
							t/data/test2.rdf
							t/data/test2.xml
							t/data/test_go.obo
							t/data/test_go_transitive_closure.obo
							t/data/test_go_transitive_reduction.obo
							t/data/test_goa.rdf
							t/data/test_goa_parser_out.obo
							t/data/test_intact_parser_out.obo
							t/data/test_ncbi.rdf
							t/data/test_ncbi_parser_out.obo
							t/data/test_ulo_apo.dot
							t/data/test_ulo_apo.owl
							t/data/test_ulo_apo2.obo
							t/data/test_uniprot_parser_out.obo
							t/data/test_uo.rdf"},
);
